File: src/main/java/com/github/squirrelgrip/meter/service/DefaultErrorHandler.java
package com.github.squirrelgrip.meter.service;

import com.github.squirrelgrip.meter.domain.ContextKey;
import com.github.squirrelgrip.meter.exception.MeterReaderException;
import com.github.squirrelgrip.meter.exception.UnknownException;

public class DefaultErrorHandler implements ErrorHandler {
    @Override
    public void handleException(MeterReaderException exception) {
        System.out.println(exception.getMessage() + " occurred at line " + exception.getContext().get(ContextKey.LINE_NUMBER));
        if (exception instanceof UnknownException) {
            System.out.println("Cause: ");
            exception.getCause().printStackTrace();
        }
    }
}

File: src/main/kotlin/com/github/squirrelgrip/meter/service/DefaultErrorHandler.kt
package com.github.squirrelgrip.meter.service

import com.github.squirrelgrip.meter.exception.MeterReaderException
import com.github.squirrelgrip.meter.exception.UnknownException

class DefaultErrorHandler : ErrorHandler {
    override fun handleException(exception: MeterReaderException?) {
        if (exception == null) {return}
        val session = exception.context.session
        if (session.isTransactionActive) {
            session.rollback()
        }
        if (exception is UnknownException) {
            println(exception.message)
            println("Cause: " + exception.cause?.message)
        } else {
            println("Error occurred at line ${exception.context.lineNumber} in ${exception.context.fileName}")
            println(exception.message)
        }
    }
}
